File: 백준/Gold/2024. 선분 덮기/선분 덮기.cc
#include <bits/stdc++.h>
using namespace std;
#define fastio                        \
    ios_base::sync_with_stdio(false); \
    cin.tie(NULL)
#define ll long long

int main()
{
    fastio;

    int m;
    cin >> m;
    vector<pair<int, int>> arr;
    while (1)
    {
        int a, b;
        cin >> a >> b;
        if (a == 0 && b == 0)
            break;
        if (b <= a)
            continue;
        arr.push_back({a, -b});
    }
    sort(arr.begin(), arr.end());
    int st = 0;
    int cnt = 0;
    auto iter = arr.begin();
    priority_queue<int> pq;
    while (st < m)
    {
        int max_end = st;
        while (iter != arr.end() && iter->first <= st)
        {
            max_end = max(max_end, -iter->second);
            iter++;
        }
        if (max_end == st)
        {
            cout << 0;
            return 0;
        }
        st = max_end;
        cnt++;
    }
    cout << cnt;
    return 0;
}

File: 백준/Gold/2024. 선분 덮기/선분 덮기.py
import heapq
m = int(input())
arr = []
while 1:
    a, b = map(int, input().split())
    if a == 0 and b == 0:
        break
    heapq.heappush(arr, [a, -b])


st = 0
cnt = 0
while st < m :
    max_end = st
    while arr and arr[0][0] <= st:
        start, end = heapq.heappop(arr)
        max_end = max(max_end, -end)        
        
    if(max_end == st):
        print(0)
        exit()
    st = max_end
    cnt += 1
print(cnt)